lib/sources.nix

```diff
@@ -101,4 +101,201 @@ rec {
+      #     trace: myproject: include regular   /home/user/myproject/images/image.svg
+      #     trace: myproject: include directory /home/user/myproject/src
+      #     trace: myproject: include directory /home/user/myproject/src/MyDir
+      #     trace: myproject: include regular   /home/user/myproject/src/MyDir/File1.hs
+      #     trace: myproject: include regular   /home/user/myproject/src/MyDir/File2.hs
+      debugEnableTracing ? false,
+      # Set this to prefix the trace output with some arbitrary string.
+      # Useful if you enable `debugEnableTracing` in multiple places and want
+      # to distinguish them.
+      debugTracePrefix ? "",
+    }: with lib;
+      let
+        # Pre-processing done once, across all files passed in.
+
+        # Turns a list into a "set" (map where all values are null).
+        keySet = list: builtins.listToAttrs (map (name: nameValuePair name null) list);
+
+        # For fast non-O(n) lookup, we turn `srcDirs` and `srcFiles` into
+        # string-keyed attrsets first.
+        srcDirsSet = keySet (map toString srcDirs);
+        srcFilesSet = keySet (map toString srcFiles);
+
+        # We also turn `srcDirs` into a directory-prefix-tree so that we can
+        # check whether a given path is under one of the `srcDirs` in sub-O(n).
+        srcDirsTree = pathComponentsToTree (map splitPath srcDirs);
+      in
+        # The actual filter function with per-file processing
+        fullPath: type:
+          let
+            fileName = baseNameOf (toString fullPath);
+
+            components = splitPath fullPath;
+
+            isExplicitSrcFile = builtins.hasAttr fullPath srcFilesSet;
+            isExplicitSrcDir = type == "directory" && builtins.hasAttr fullPath srcDirsSet;
+            # The below is equivalent to
+            #   any (srcDir: hasPrefix (toString srcDir + "/") fullPath) srcDirs;
+            # but faster than O(n) where n is the number of `srcDirs` entries.
+            isUnderSomeSrcDir = isPrefixOfLeafPath components srcDirsTree;
+
+            isHidden = excludeHidden && hasPrefix "." fileName;
+
+            hasExcludedComponentInPath = any (c: elem c pathComponentExcludes) components;
+
+            isSourceInput =
+              isExplicitSrcFile ||
+              isExplicitSrcDir ||
+              (isUnderSomeSrcDir && !isHidden && !hasExcludedComponentInPath);
+
+            tracing =
+              let
+                prefix = if debugTracePrefix == "" then "" else debugTracePrefix + ": ";
+                action = if isSourceInput then "include" else "skip   ";
+                # Pad "regular" to be as wide as "directory" for aligned output.
+                formattedType = if type == "regular" then "regular  " else type;
+              in
+              if debugEnableTracing
+                then builtins.trace "${prefix}${action} ${formattedType} ${fullPath}"
+                else id;
+          in
+            tracing isSourceInput;
```

Comment:
You could use the newish `lib.debug.traceIf` function here.

Reply:
Done.